src/InsertCapablePdoTrait.php

```diff
@@ -0,0 +1,236 @@
+<?php
+
+namespace RebelCode\Storage\Resource\Pdo;
+
+use ArrayAccess;
+use Dhii\Util\String\StringableInterface;
+use OutOfRangeException;
+use Psr\Container\ContainerExceptionInterface;
+use Psr\Container\NotFoundExceptionInterface;
+use Dhii\Util\String\StringableInterface as Stringable;
+use InvalidArgumentException;
+use PDOStatement;
+use Psr\Container\ContainerInterface;
+use stdClass;
+use Traversable;
+
+/**
+ * Common functionality for objects that can insert records into a database using PDO.
+ *
+ * @since [*next-version*]
+ */
+trait InsertCapablePdoTrait
+{
+    /**
+     * Executes an INSERT SQL query, inserting records into the database.
+     *
+     * @since [*next-version*]
+     *
+     * @param array[]|ArrayAccess[]|stdClass[]|ContainerInterface[]|Traversable $records A list of records to insert.
+     *
+     * @throws ContainerExceptionInterface If an error occurred while reading from a record's container.
+     *
+     * @return PDOStatement The executed PDO statement.
+     */
+    protected function _insert($records)
+    {
+        $processedRecords = $this->_preProcessRecords($records, $valueHashMap);
+
+        $query = $this->_buildInsertSql(
+            $this->_getSqlInsertTable(),
+            $this->_getSqlInsertColumnNames(),
+            $processedRecords,
+            $valueHashMap
+        );
+
+        $statement = $this->_executePdoQuery($query, array_flip($valueHashMap));
+
+        return $statement;
+    }
+
+    /**
+     * Pre-processes the list of records.
+     *
+     * @since [*next-version*]
+     *
+     * @param array[]|ArrayAccess[]|stdClass[]|ContainerInterface[]|Traversable $records      A list of records.
+     * @param array                                                             $valueHashMap A hash-to-value map
+     *                                                                                        reference to which new
+     *                                                                                        hash-value pairs are
+     *                                                                                        written.
+     *
+     * @throws ContainerExceptionInterface If an error occurred while reading from a record's container.
+     *
+     * @return array The pre-processed record data list, as an array of record data associative sub-arrays.
+     */
+    protected function _preProcessRecords($records, &$valueHashMap = [])
```

Comment:
Since this specifically converts key-value mappings into key-valhash mappings, does it make sense for this method to have such a generic name? This name hints to me that it maybe normalizes some values, not that it entirely changes them - inserting value hashes into a table wouldn't make much sense.

Reply:
Pretty much what I said [here](https://github.com/RebelCode/pdo-cqrs-resource-models/pull/1#discussion_r171797876). Any suggestions for the name?

Can you clarify what you meant by "inserting value hashes into a table wouldn't make much sense."? Is it because you disagree or don't understand why values are changed into hashes?

Reply:
I understand why they are changed into hashes. But the typical usage algo goes like this:

Reply:
I understand. How about something a bit simpler than "surrogate"? Like "replace" or "substitute"? Currently leaning towards `_substituteRecordValuesForHashes()`.